app/js/app.js

```diff
@@ -68,9 +68,54 @@ if (console) {
               "font-size:1.5em;color:#4558c9;", "color:#d61a7f;font-size:1em;");
 }
 
+// Show a toast with a service-worker-related update.
+window.showToast = function(message) {
+  var toast = document.getElementById('sw-toast');
+
+  if (!toast) {
+    toast = document.createElement('paper-toast');
+    toast.id = 'sw-toast';
+    document.body.appendChild(toast);
+  }
+
+  toast.textContent = message;
+  toast.show();
+};
+
 // Register service worker if supported.
 if ('serviceWorker' in navigator) {
-  navigator.serviceWorker.register('/service-worker.js');
+  navigator.serviceWorker.register('/service-worker.js'
+  ).then(function(registration) {
+    registration.onupdatefound = function() {
+      // The updatefound event implies that registration.installing is set; see
+      // https://slightlyoff.github.io/ServiceWorker/spec/service_worker/index.html#service-worker-container-updatefound-event
+      const installingWorker = registration.installing;
+      installingWorker.onstatechange = function() {
+        switch (installingWorker.state) {
+          case 'installed':
+            if (!navigator.serviceWorker.controller) {
+              window.showToast('Caching complete! Future visits will work offline.');
+            }
+            break;
+
+          case 'redundant':
+            throw Error('The installing service worker became redundant.');
+        }
+      };
+    };
+  }).catch(function(e) {
+    console.error('Service worker registration failed:', e);
+  });
+}
+
+// Check to see if the service worker controlling the page at initial load
+// has become redundant, since this implies there's a new service worker with fresh content.
+if (navigator.serviceWorker && navigator.serviceWorker.controller) {
+  navigator.serviceWorker.controller.onstatechange = function(event) {
+    if (event.target.state === 'redundant') {
+      window.showToast('A new version of this app is available. Refresh the page.');
```

Comment:
Rather than the app-y language, maybe just, "Site updated. Refresh this page to see the latest content." ? May overpromise since the current page may _have_ no new content. But still... WDYT?